### PR TITLE
Adding LastTransitionTime to PodCondition

### PR DESCRIPTION
**Pull request by nikhiljindal**
Fixes https://github.com/kubernetes/kubernetes/issues/12615
Required for https://github.com/kubernetes/kubernetes/issues/1743
Enhancing PodCondition by adding LastTransitionTime, Reason and Message.

cc @bgrant0607 for API changes and 
@dchen1107 @yujuhong for kubelet changes

### CODE REVIEW

**Comment by bgrant0607** [comment]
cc @mikedanese and/or @davidopp on first-pass review of API changes

**Comment by davidopp** [comment]
LGTM

(I checked the whole thing, excluding the test)

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -2277,29 +2277,66 @@ func GetPhase(spec *api.PodSpec, info []api.ContainerStatus) api.PodPhase {
...
-	unready := []api.PodCondition{{
-		Type:   api.PodReady,
-		Status: api.ConditionFalse,
-	}}
-	if statuses == nil {
-		return unready
+func getPodReadyCondition(spec *api.PodSpec, containerStatuses []api.ContainerStatus, existingStatus *api.PodStatus) []api.PodCondition {
+	// Find if all containers are ready or not.
+	if containerStatuses == nil {
+		return readyPodCondition(false, "Could not get container statuses", "", existingStatus)
```
Comment (dchen1107) [inline pkg/kubelet/kubelet.go]: Shouldn't reason be a short and brief? "Could not get container statuses" should be message, not reason.

File: pkg/api/v1/types.go [hunk 2]
```diff
@@ -826,12 +826,12 @@ const (
...
 type PodCondition struct {
-	// Type is the type of the condition
-	Type PodConditionType `json:"type" description:"kind of the condition, currently only Ready; see http://releases.k8s.io/HEAD/docs/user-guide/pod-states.md#pod-conditions"`
-	// Status is the status of the condition
-	Status ConditionStatus `json:"status" description:"status of the condition, one of True, False, Unknown; see http://releases.k8s.io/HEAD/docs/user-guide/pod-states.md#pod-conditions"`
+	Type               PodConditionType `json:"type" description:"kind of the condition, currently only Ready; see http://releases.k8s.io/HEAD/docs/user-guide/pod-states.md#pod-conditions"`
+	Status             ConditionStatus  `json:"status" description:"status of the condition, one of True, False, Unknown; see http://releases.k8s.io/HEAD/docs/user-guide/pod-states.md#pod-conditions"`
+	LastTransitionTime util.Time        `json:"lastTransitionTime,omitempty" description:"last time the condition transitioned from one status to another"`
+	Reason             string           `json:"reason,omitempty" description:"(brief) reason for the condition's last transition"`
```
Comment (dchen1107) [inline pkg/api/v1/types.go]: s/(brief)/(brief-CamelCase) here to make all reasons consistent?

File: pkg/kubelet/kubelet.go [hunk 3]
```diff
@@ -2277,29 +2277,66 @@ func GetPhase(spec *api.PodSpec, info []api.ContainerStatus) api.PodPhase {
...
+	if len(containersWithUnreadyStatus) > 0 {
+		if unreadyMessage == "" {
+			unreadyMessage = fmt.Sprintf("Containers with unready status: %s", containersWithUnreadyStatus)
+		} else {
+			unreadyMessage = fmt.Sprintf("%s, containers with unready status: %s", unreadyMessage, containersWithUnreadyStatus)
+		}
+	}
+	if unreadyMessage != "" {
+		// return unready status.
+		return readyPodCondition(false, fmt.Sprintf("Some containers are not ready"), unreadyMessage, existingStatus)
```
Comment (dchen1107) [inline pkg/kubelet/kubelet.go]: Reason could be ContainerNotReady, brief-CamelCase :-)

**Comment by dchen1107** [comment]
Overall LGTM except above minor issues. But I do expect you fix them since we expect Reason to be a concise, single-word, CamelCase. @bgrant0607 mentioned this to me a while back on cleaning up Kubelet reporting Reason, we just haven't get around to do it yet. But for new code, they want them to be clean at the initial time. Thanks!

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (nikhiljindal) [inline pkg/kubelet/kubelet.go]: Done

File: pkg/api/v1/types.go (hunk 2 above)
Comment (nikhiljindal) [inline pkg/api/v1/types.go]: Done

File: pkg/kubelet/kubelet.go (hunk 3 above)
Comment (nikhiljindal) [inline pkg/kubelet/kubelet.go]: Done

**Comment by nikhiljindal** [comment]
Thanks for your comments @dchen1107 
Have pushed new code with the fixes.
Thanks!

**Comment by bgrant0607** [comment]
Thanks for catching the misuse of Reason @dchen1107!

@nikhiljindal Please read: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#typical-status-properties

File: pkg/api/v1/types.go [hunk 4]
```diff
@@ -826,12 +826,12 @@ const (
...
 type PodCondition struct {
-	// Type is the type of the condition
-	Type PodConditionType `json:"type" description:"kind of the condition, currently only Ready; see http://releases.k8s.io/HEAD/docs/user-guide/pod-states.md#pod-conditions"`
-	// Status is the status of the condition
-	Status ConditionStatus `json:"status" description:"status of the condition, one of True, False, Unknown; see http://releases.k8s.io/HEAD/docs/user-guide/pod-states.md#pod-conditions"`
+	Type               PodConditionType `json:"type" description:"kind of the condition, currently only Ready; see http://releases.k8s.io/HEAD/docs/user-guide/pod-states.md#pod-conditions"`
+	Status             ConditionStatus  `json:"status" description:"status of the condition, one of True, False, Unknown; see http://releases.k8s.io/HEAD/docs/user-guide/pod-states.md#pod-conditions"`
+	LastTransitionTime util.Time        `json:"lastTransitionTime,omitempty" description:"last time the condition transitioned from one status to another"`
+	Reason             string           `json:"reason,omitempty" description:"(brief-CamelCase) reason for the condition's last transition"`
```
Comment (bgrant0607) [inline pkg/api/v1/types.go]: unique, one-word, CamelCase reason

File: pkg/api/v1/types.go [hunk 5]
```diff
@@ -826,12 +826,12 @@ const (
...
+	Message            string           `json:"message,omitempty" description:"human readable message indicating details about last transition"`
```
Comment (bgrant0607) [inline pkg/api/v1/types.go]: human-readable

**Comment by bgrant0607** [comment]
We should also change kubectl get pods to use this reason in the STATUS column in the case that it's non-empty and the Ready condition status is False.
cc @JanetKuo

**Comment by dchen1107** [comment]
@nikhiljindal once you update these two field descriptions based on https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#typical-status-properties

Please ping me, and we are ready to go. Thanks!

File: pkg/api/v1/types.go (hunk 4 above)
Comment (nikhiljindal) [inline pkg/api/v1/types.go]: Done

File: pkg/api/v1/types.go (hunk 5 above)
Comment (nikhiljindal) [inline pkg/api/v1/types.go]: Done

**Comment by nikhiljindal** [comment]
Thanks @bgrant0607  and @dchen1107 . Updated code as per comments

**Comment by dchen1107** [comment]
LGTM from Dawn for real. :-)

**Comment by mikedanese** [comment]
@brendandburns k8s-merge-bot is incorrectly removing lgtm's

**Comment by dchen1107** [comment]
LGTM again

**Comment by bgrant0607** [comment]
I want to hold off on this until #10933 is merged.

**Comment by nikhiljindal** [comment]
Now that #10933 is merged, this is now ready!
@bgrant0607 Can I haz merge now?

**Comment by bgrant0607** [comment]
Please rebase and convert the description tags to comments.

**Comment by bgrant0607** [comment]
Sorry, didn't see your update -- github didn't refresh.

Did you drop LastHeartbeatTime?

**Comment by bgrant0607** [comment]
Both shippable runs failed. Restart if you think they were flakes.

**Comment by nikhiljindal** [comment]
Restarted shippable.
I never added LastHearbeatTime.

**Comment by bgrant0607** [comment]
Let's go ahead and make this symmetric with NodeCondition. Unfortunately, "Heartbeat" is not quite appropriate, since it implies upward communication. So I think we should call it LastProbeTime.

**Comment by nikhiljindal** [comment]
@bgrant0607 Have added LastProbeTime.
Please take another look!

File: pkg/kubelet/kubelet_test.go [hunk 6]
```diff
@@ -1671,30 +1671,47 @@ func getNotReadyStatus(cName string) api.ContainerStatus {
 		Ready: false,
 	}
 }
+func getReadyCondition(status api.ConditionStatus, transitionTime util.Time, reason, message string) []api.PodCondition {
+	return []api.PodCondition{{
+		Type:               api.PodReady,
+		Status:             status,
+		LastTransitionTime: transitionTime,
```
Comment (bgrant0607) [inline pkg/kubelet/kubelet_test.go]: Add LastProbeTime.

Comment (nikhiljindal) [inline pkg/kubelet/kubelet_test.go]: We are not comparing the ProbeTime in tests, since we cant inject or override time.Now().

One way to fix that would be to add a setNowForTest() method to util.Time and update Now() to return that time instead of the current time, if it is set. Then we can check that thats the time set in LastProbeTime.
But I wasnt sure if thats a good way. WDYT?

Comment (nikhiljindal) [inline pkg/kubelet/kubelet_test.go]: cc @lavalamp 
Based on irl discussion with @bgrant0607 we need a way to test fields based on util.Time.
Filed https://github.com/kubernetes/kubernetes/issues/13399 to fix that.

**Comment by bgrant0607** [comment]
I'm fine with the API change. Reassigning for code review.

File: pkg/kubelet/kubelet.go [hunk 7]
```diff
@@ -2316,29 +2316,68 @@ func GetPhase(spec *api.PodSpec, info []api.ContainerStatus) api.PodPhase {
...
+	if len(containersWithUnreadyStatus) > 0 {
+		if unreadyMessage == "" {
+			unreadyMessage = fmt.Sprintf("Containers with unready status: %s", containersWithUnreadyStatus)
+		} else {
+			unreadyMessage = fmt.Sprintf("%s, containers with unready status: %s", unreadyMessage, containersWithUnreadyStatus)
+		}
+	}
+	if unreadyMessage != "" {
+		// return unready status.
+		return readyPodCondition(false, fmt.Sprintf("ContainersNotReady"), unreadyMessage, existingStatus)
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Don't think you need fmt.Sprintf here

File: pkg/kubelet/kubelet.go [hunk 8]
```diff
@@ -2316,29 +2316,68 @@ func GetPhase(spec *api.PodSpec, info []api.ContainerStatus) api.PodPhase {
...
 		}
 	}
-	return ready
+	unreadyMessage := ""
+	if len(containersWithUnknownStatus) > 0 {
+		unreadyMessage = fmt.Sprintf("Containers with unknown status: %s", containersWithUnknownStatus)
+	}
+	if len(containersWithUnreadyStatus) > 0 {
+		if unreadyMessage == "" {
+			unreadyMessage = fmt.Sprintf("Containers with unready status: %s", containersWithUnreadyStatus)
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Just a nit (and you are free to ignore it). 

If capitalization is not important here, I'd suggest creating a []string, append it, and join them in the end to create the unready message.

File: pkg/kubelet/kubelet.go [hunk 9]
```diff
@@ -2316,29 +2316,68 @@ func GetPhase(spec *api.PodSpec, info []api.ContainerStatus) api.PodPhase {
...
-	}}
-	if statuses == nil {
-		return unready
+func getPodReadyCondition(spec *api.PodSpec, containerStatuses []api.ContainerStatus, existingStatus *api.PodStatus) []api.PodCondition {
+	// Find if all containers are ready or not.
+	if containerStatuses == nil {
+		return readyPodCondition(false, "UnknownContainerStatuses", "", existingStatus)
 	}
+	containersWithUnknownStatus := []string{}
+	containersWithUnreadyStatus := []string{}
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Just a nit.

I think shorter variable names here would be fine since this is not a huge function, and it's quite clearly what the code intends to do. How about  "unreadyContainers", "unknownContainers"?

File: pkg/kubelet/kubelet.go [hunk 10]
```diff
@@ -2393,8 +2432,12 @@ func (kl *Kubelet) generatePodStatus(pod *api.Pod) (api.PodStatus, error) {
 			}
 		}
 	}
-
-	podStatus.Conditions = append(podStatus.Conditions, getPodReadyCondition(spec, podStatus.ContainerStatuses)...)
+	existingStatus, ok := kl.statusManager.GetPodStatus(pod.UID)
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: You may want to rebase because this function is slightly behind. GetPodStatus is already called in this function at HEAD.

**Comment by lavalamp** [comment]
+1 to all of @yujuhong's nits

File: pkg/kubelet/kubelet.go (hunk 7 above)
Comment (nikhiljindal) [inline pkg/kubelet/kubelet.go]: Yes changed to Sprint

File: pkg/kubelet/kubelet.go (hunk 8 above)
Comment (nikhiljindal) [inline pkg/kubelet/kubelet.go]: Done

File: pkg/kubelet/kubelet.go (hunk 9 above)
Comment (nikhiljindal) [inline pkg/kubelet/kubelet.go]: Changed to shorter names

File: pkg/kubelet/kubelet.go (hunk 10 above)
Comment (nikhiljindal) [inline pkg/kubelet/kubelet.go]: Done

**Comment by nikhiljindal** [comment]
Thanks @yujuhong for the review. Have updated code as per comments.
PTAL.

**Comment by yujuhong** [comment]
@k8s-bot ok to test

**Comment by yujuhong** [comment]
LGTM, but shippable and jenkins are not happy.

**Comment by quinton-hoole** [comment]
There is so much cruft in the integration test logs that it's really difficult to figure out exactly what failed.  It could be this, but I'm not sure:

```
F0901 23:35:59.316351   14760 integration.go:856] FAILED: pod never started running timed out waiting for the condition
!!! Error in ./hack/test-integration.sh:51
  'KUBE_API_VERSIONS="v1" "${KUBE_OUTPUT_HOSTBIN}/integration" --v=${LOG_LEVEL} --api-version="$1" --max-concurrency="${KUBE_INTEGRATION_TEST_MAX_CONCURRENCY}"' exited with status 255
```
Anyhow, I restarted the Shippable build.

The two Services tests that failed seem to be pretty flaky too . I'm still trying to get the the bottom of that in my copious spare time :-)  Lets retry that in the mean time too...

@k8s-bot test this again please

**Comment by quinton-hoole** [comment]
Nope, shippable failed again. @nikhiljindal could you take a look?

**Comment by quinton-hoole** [comment]
Assigning to @nikhiljindal to address the shippable failures.

**Comment by bgrant0607** [comment]
Ping. Please look at test failures.

**Comment by nikhiljindal** [comment]
I cant see why the integration test is failing.
@yujuhong anything obvious that I maybe missing?

**Comment by yujuhong** [comment]
> I cant see why the integration test is failing.
> @yujuhong anything obvious that I maybe missing?

You are simply unlucky and is hitting a corner case, where kubelet is busy dealing with constant, false update caused by annotation difference...apply the fix in #13880 (waiting to be merged, soon), and the tests should pass.

**Comment by nikhiljindal** [comment]
Thanks a lot @yujuhong. You are awesome!

I rebased after that PR got merged and my tests pass now!!

**Comment by nikhiljindal** [comment]
Putting my LGTM back since there is no code change, just sync and rebase.

**Comment by nikhiljindal** [comment]
Sigh!
@k8s-bot ok to test

**Comment by jszczepkowski** [comment]
@k8s-bot test this

**Comment by nikhiljindal** [comment]
Had to rebase since util.Time got moved to unversioned.Time
Tests should pass now.

File: pkg/kubelet/kubelet.go [hunk 11]
```diff
@@ -2126,29 +2126,65 @@ func GetPhase(spec *api.PodSpec, info []api.ContainerStatus) api.PodPhase {
...
+	currentTime := unversioned.Now()
+	condition := api.PodCondition{
+		Type: api.PodReady,
 	}
+	if isPodReady {
+		condition.Status = api.ConditionTrue
+	} else {
+		condition.Status = api.ConditionFalse
+	}
+	condition.LastProbeTime = currentTime
```
Comment (lavalamp) [inline pkg/kubelet/kubelet.go]: This causes kubelet to send drastically more status messages! https://github.com/kubernetes/kubernetes/issues/14273#issuecomment-142451861
